iceoryx_hoofs/include/iceoryx_hoofs/internal/relocatable_pointer/relative_pointer.hpp

```diff
@@ -82,9 +82,19 @@ class RelativePointer : public BaseRelativePointer
     /// @return a pointer to the underlying object
     T* get() const noexcept;
 
-    /// @brief converts the RelativePointer to a pointer of the type of the underlying object
-    /// @return a pointer of type T pointing to the underlying object
-    operator T*() const noexcept;
+    /// @brief converts the RelativePointer to bool
+    /// @return bool which contains true if the RelativePointer contains a pointer
+    explicit operator bool() const noexcept;
+
+    /// @brief True if the underyling pointer is smaller than the one in the rhs
+    /// @param[in] rhs Another RelativePointer
+    /// @return True if smaller, otherwise false
+    bool operator<(const RelativePointer<T>& rhs) const noexcept;
```

Comment:
Do we need these operators? Usually pointer comparison like this is a bad idea. Equality comparison is useful, but order not so much. One reason is that a order has to satisfy certain properties (basically be at least a irreflexive transitive relation (this is implicitly assumed by various STL constructs).

E.g. `std::map<K,V>` requires that these operators exist for K, but I am not aware that we use these pointers in a map. If the relation < exists but does not have these properties, weird things happen.

One can define it to essentially be the < relation on the corresponding raw pointer.
However, it makes little sense to compare these kinds of pointers as the relation cannot hold across process boundaries (i.e. in one process the pointers may compare as <, in the other as >). This may lead to all kinds of unpredictable problems.

Implicitly this was the case before as well, but not really responsibility of the class.

TLDR: Expecting an order on raw pointers is a bad idea. Expecting an order on fancy pointers is worse (especially those that have to work across process boundaries). The operators should be removed unless really needed. If the user wants to compare the raw pointers after a `get` she should do just that (and be aware of the problems).

Reply:
We store `RelativePointer`s in `cxx::vector`. The operators are required for `std::sort` in file `iceoryx_posh/internal/popo/building_blocks/chunk_distributor.inl:179-180`. The code was added for the blocking publisher feature. I understand your point and I'm not sure if we uncovered a conversion problem here. My understanding:

```
// chunk_distributor_data.hpp
using QueueContainer_t =
    cxx::vector<rp::RelativePointer<ChunkQueueData_t>, ChunkDistributorDataProperties_t::MAX_QUEUES>;

// chunk_distributor.inl
std::sort(getMembers()->m_queues.begin(), getMembers()->m_queues.end());

// vector.inl
using const_iterator = const T*; // with T being RelativePointer
return reinterpret_cast<const_iterator>(&at_unchecked(0));

This used to call the conversion operator and use the `operator<(const ChunkQueueData& lhs, ChunkQueueData rhs)`. This is now done via the new operators explicitly.
```

Reply:
The question is what should be sorted here? If its is just to have a well-defined order it is fine to sort by pointer. In fact I think sorting by data here is not well-defined anyway.

The elements point to something in shared memory, which in a fixed process address space can act as unique identifiers and are sortable by  pointer. I just think that the solution is to use an explicit sorting criterion with `std::sort`, e.g. with a functor struct or function that defines the order.

Reply:
@elfenpiff Can you answer the above question? Looking again at the code, I think we want to sort the pointers, correct?

Reply:
Talked to @elfenpiff the other day, the sorting mechanism is indeed wanted here and correct.